packages/block-library/src/table-of-contents/index.php

```diff
@@ -0,0 +1,174 @@
+			)
+		);
+	}
+
+	register_block_type_from_metadata(
+		__DIR__ . '/table-of-contents',
+		array(
+			'render_callback' => 'render_block_core_table_of_contents',
+		)
+	);
+}
+add_action( 'init', 'register_block_core_table_of_contents' );
+
+/**
+ * Takes a flat list of heading parameters and nests them based on each header's
+ * immediate parent's level.
+ *
+ * @since 6.4.0
+ *
+ * @param array $headings A flat list of headings to nest.
+ * @return array $tree An array of nested headings.
+ */
+function block_core_table_of_contents_build_headings_tree( $headings ) {
+	if ( empty( $headings ) ) {
+		return $headings;
+	}
+
+	$tree          = array();
+	$first_heading = $headings[0];
+	$total         = count( $headings );
+
+	foreach ( $headings as $key => $heading ) {
+		if ( empty( $heading['content'] ) ) {
+			continue;
+		}
+
+		// Make sure we are only working with the same level as the first iteration in our set.
+		if ( $heading['level'] !== $first_heading['level'] ) {
+			continue;
+		}
+
+		/**
+		 * Check that the next iteration will return a value.
+		 * If it does and the next level is greater than the current level,
+		 * the next iteration becomes a child of the current iteration.
+		 */
+		$offset = $key + 1;
+		if ( isset( $headings[ $offset ]['level'] ) && (int) $headings[ $offset ]['level'] > (int) $heading['level'] ) {
+			/**
+			 * Calculate the last index before the next iteration with the same level (siblings).
+			 * Then, use this index to slice the array for use in recursion.
+			 * This prevents duplicate nodes.
+			 */
+			$end = $total;
+			for ( $i = $offset; $i < $total; $i++ ) {
+				if ( $headings[ $i ]['level'] === $heading['level'] ) {
+					$end = $i;
+					break;
+				}
+			}
+
+			$tree[] = array(
+				'heading'  => $heading,
+				'children' => block_core_table_of_contents_build_headings_tree( array_slice( $headings, $offset, $end - 1 ) ),
+			);
+
+		} else {
+			// No child heading levels.
+			$tree[] = array(
+				'heading' => $heading,
+			);
+		}
+	}
+
+	return $tree;
+}
+
+/**
+ * Build a list of Table of Content items.
+ *
+ * @since 6.4.0
+ *
+ * @param array $tree An array of nested headings.
+ * @return string $list A list of Table of Content items.
+ */
+function block_core_table_of_contents_build_list( $tree ) {
+	$list      = '';
+	$permalink = get_permalink();
+
+	foreach ( $tree as $item ) {
+		$heading  = $item['heading'];
+		$children = isset( $item['children'] ) ? '<ol>' . block_core_table_of_contents_build_list( $item['children'] ) . '</ol>' : '';
+
+		if ( ! empty( $heading['link'] ) ) {
+			$pagelink = ! empty( $heading['page'] ) ? add_query_arg( 'page', $heading['page'], $permalink ) : $permalink;
+			$content  = '<a class="wp-block-table-of-contents__entry" href="' . esc_url( $pagelink . $heading['link'] ) . '">' . esc_html( $heading['content'] ) . '</a>';
+		} else {
+			$content = '<span class=wp-block-table-of-contents__entry>' . esc_html( $heading['content'] ) . '</span>';
```

Comment:
Nit: Ideally, we should keep the quotes around each attribute.

```suggestion
			$content = '<span class="wp-block-table-of-contents__entry">' . esc_html( $heading['content'] ) . '</span>';
```